File: HW3/hw3_problem1.m
%% MAE 6760 Model Based Estimation
% Cornell University
% M Campbell
%
% Homework #3
% Problem #1: Kalman Filter
%   Five mass-spring-damper system with initial condition and disturbance
%
clear all; close all
MAE6760startup; %can adjust font size, figure size at the bottom of this script

%% Define system model: continuous and then discrete
k1=1;k2=1;k3=1;k4=1;k5=1;
c1=0.01;c2=0.01;c3=0.01;c4=0.01;c5=0.01;
m1=1;m2=1;m3=1;m4=1;m5=1;
%System matrix, with states q1,q1dot,...
nx=10;
Fc=zeros(nx,nx);
Fc([2 4 6 8 10],[1 3 5 7 9])=...
 [-(k1+k2)/m1 (k2)/m1 0 0 0; 
   (k2)/m2 -(k2+k3)/m2 (k3)/m2 0 0;
   0 (k3)/m3 -(k3+k4)/m3 (k4)/m3 0;
   0 0 (k4)/m4 -(k4+k5)/m4 (k5)/m4;
   0 0 0 (k5)/m5 -(k5)/m5];      
Fc([2 4 6 8 10],[2 4 6 8 10])=...
 [-(c1+c2)/m1 (c2)/m1 0 0 0; 
   (c2)/m2 -(c2+k3)/m2 (c3)/m2 0 0;
   0 (c3)/m3 -(c3+k4)/m3 (c4)/m3 0;
   0 0 (c4)/m4 -(c4+c5)/m4 (c5)/m4;
   0 0 0 (c5)/m5 -(c5)/m5]; 
Fc([1 3 5 7 9],[2 4 6 8 10])=eye(5);
%disturbance w on the 5th mass
nw=1;
Gc=zeros(nx,1);Gc(nx,1)=1/m5;
nz=2;
Hc=[1 0 zeros(1,8); %position of the 1st mass
    zeros(1,8) 1 0];%position of the 5th mass
%CT model
CTsys=ss(Fc,Gc,Hc,zeros(nz,nw));
%DT model
dt=0.05;
DTsys = c2d(CTsys,dt);

%% Simulate the dynamic system
%initial condition on the position 5th mass
x0=zeros(nx,1);x0(9)=1; %displace the 5th mass
t=[0:dt:50];nk=length(t);
rng(101) %sets the same random number generator seed
Qsim=0.01;
w=randn(nk,1)*sqrt(Qsim);
[~,tsim,x_no_w] = lsim(DTsys,w*0,t,x0); %response to just the initial condition
x_no_w=x_no_w'; %make into row vector
[~,tsim,x_true] = lsim(DTsys,w,t,x0);   %response to the initial condition and disturbance
x_true=x_true'; %make into row vector

%Create noisy measurements
i1=1;
Rq1=0.001;vq1=sqrt(Rq1)*randn(1,nk);
z_q1=x_true(i1,:)+vq1;
i5=9;
Rq5=0.001;vq5=sqrt(Rq5)*randn(1,nk);
z_q5=x_true(i5,:)+vq5;

%plots of the states and measurements for the 1st/5th masses
figs(1)=figure;
ti(1)=tiledlayout(1,2,'TileSpacing','compact','Padding','tight');
nexttile;
i1=1;
plot_openloop(t,x_true(i1,:),z_q1,x_no_w(i1,:))
ylabel('position of the 1st mass');xlabel('time (sec)');
nexttile;
i2=9;
plot_openloop(t,x_true(i2,:),z_q5,x_no_w(i2,:))
ylabel('position of the 5th mass');xlabel('time (sec)');
title(ti(1),'simulated response: 1st and 5th masses','fontweight','bold');


%% Part (a): two Kalman Filters using z_q1, z_q5
%KF using z_q1
F=DTsys.A;G=DTsys.B;Q=Qsim;
H=DTsys.C(1,:);
R=Rq1;z=z_q1;

%%YOUR CODE HERE

%Generate plot of position error estimates for the 1st and 5th masses and 2-sigma bounds
%assumes t is (1 x nk), xhatu is (nx x nk), Pu is (nx x nx x nk), x_true is (nx x nk) 
%measurement of position of 1st mass
figs(2)=figure;
ti(2)=tiledlayout(1,2,'TileSpacing','compact','Padding','tight');
nexttile;
plot_estimator(t,xhatu(i1,:),Pu(i1,i1,:),x_true(i1,:),'error',z);
ylabel('error position q1');
ylim([-0.3 0.3]);
nexttile;
plot_estimator(t,xhatu(i2,:),Pu(i2,i2,:),x_true(i2,:),'error');
ylabel('error position q5');
ylim([-2 2]);
title(ti(2),'(a): measurement z_{q1}','fontweight','bold');

%KF using z_q5
F=DTsys.A;G=DTsys.B;Q=Qsim;
H=DTsys.C(2,:);
R=Rq5;z=z_q5;

%%YOUR CODE HERE

%Generate plot of position error estimates for the 1st and 5th masses and 2-sigma bounds
%assumes t is (1 x nk), xhatu is (nx x nk), Pu is (nx x nx x nk), x_true is (nx x nk) 
%measurement of position of 5th mass
figs(3)=figure;
ti(3)=tiledlayout(1,2,'TileSpacing','compact','Padding','tight');
nexttile;
plot_estimator(t,xhatu(i1,:),Pu(i1,i1,:),x_true(i1,:),'error');
ylabel('error position q1');
ylim([-1 1]);
nexttile;
plot_estimator(t,xhatu(i2,:),Pu(i2,i2,:),x_true(i2,:),'error',z);
ylabel('error position q5');
ylim([-0.3 0.3]);
title(ti(3),'(a): measurement z_{q5}','fontweight','bold');


%% Part (b): predicted vs updated covariance using z_q5
%KF using z_q5
F=DTsys.A;G=DTsys.B;Q=Qsim;
H=DTsys.C(2,:);
R=Rq5;z=z_q5;

%%YOUR CODE HERE

figs(4)=figure('Position',[100 100 800 600]);
semilogy(t,[squeeze(Pp(i5,i5,:))],'b:',t,[squeeze(Pu(i5,i5,:))],'r-');
axis([0 50 1E-5 1E-3]);
grid
legend('predicted','updated');
ylabel('estimate');xlabel('time (sec)');
title('(b) predicted vs updated');


%% Part (c): Steady State error covariance
F=DTsys.A;G=DTsys.B;Q=Qsim;
H=DTsys.C(2,:);
R=Rq5;z=z_q5;

%%YOUR CODE HERE

%simulated measurement covariance from the KF at the final time:
disp('The simulated measurement covariance at the final time is:')
disp([Pz_sim]);

%steady state measurement covariance:
disp('The steady state measurement covariance is:')
disp([Pz_ss]);


%% Part (d): Kalman Filter initialized with steady state error covariance
%KF using z_q5
F=DTsys.A;G=DTsys.B;Q=Qsim;
H=DTsys.C(2,:);
R=Rq5;z=z_q5;

%%YOUR CODE HERE

%Generate plot of position error estimates for the 1st and 5th masses and 2-sigma bounds
%assumes t is (1 x nk), xhatu is (nx x nk), Pu is (nx x nx x nk), x_true is (nx x nk) 
%measurement of position of 5th mass
figs(5)=figure;
ti(5)=tiledlayout(1,2,'TileSpacing','compact','Padding','tight');
nexttile;
plot_estimator(t,xhatu(i1,:),Pu(i1,i1,:),x_true(i1,:),'error');
ylabel('error position q1');
ylim([-0.1 0.1]);
nexttile;
plot_estimator(t,xhatu(i2,:),Pu(i2,i2,:),x_true(i2,:),'error',z);
ylabel('error position q5');
ylim([-0.2 0.2]);
title(ti(5),'(d): constant KF gain with msmt z_{q5}','fontweight','bold');


%% --------- start-up items
function MAE6760startup(font_size);
%
% define colors for plotting
global MCcolors; %define colors as global with access 
MCcolors.red=[200,0,0]/255;
MCcolors.blue=[4,51,255]/255;
MCcolors.purple=[147,23,255]/255;
MCcolors.green=[0,160,0]/255;
MCcolors.orange=[253,128,8]/255;
MCcolors.mag=[255,64,255]/255;
MCcolors.cyan=[0,230,255]/255;
%
% define standard figure positioning and size
set(groot,'DefaultFigureUnits','pixels');
set(groot,'DefaultFigurePosition',[100 100 1600 600]);
set(groot,'DefaultFigureWindowStyle','normal');  % Important
set(groot,'DefaultAxesFontSize',16);
set(groot,'DefaultAxesFontWeight','bold');
set(groot,'DefaultLineLineWidth',2);
%
end

File: HW3/hw3_problem1.py
"""
MAE 6760 Model Based Estimation
Cornell University
M Campbell

Homework #3
Problem #1: Kalman Filter
    five mass–spring–damper system
    uses the following Python functions:
        plot_openloop.py
        plot_estimator.py
"""

import numpy as np
import matplotlib.pyplot as plt
from scipy.signal import cont2discrete
from scipy.linalg import solve_discrete_are

from plot_openloop import plot_openloop
from plot_estimator import plot_estimator

# ------------------------------------------------------------
# Define continuous-time system
# ------------------------------------------------------------
k1=k2=k3=k4=k5=1
c1=c2=c3=c4=c5=0.01
m1=m2=m3=m4=m5=1

nx = 10
Fc = np.zeros((nx,nx))

# stiffness
Fc[np.ix_([1,3,5,7,9],[0,2,4,6,8])] = np.array([
    [-(k1+k2)/m1,  (k2)/m1,      0,        0,      0],
    [ (k2)/m2, -(k2+k3)/m2, (k3)/m2,      0,      0],
    [ 0, (k3)/m3, -(k3+k4)/m3, (k4)/m3,   0],
    [ 0, 0, (k4)/m4, -(k4+k5)/m4, (k5)/m4],
    [ 0, 0, 0, (k5)/m5, -(k5)/m5]
])

# damping
Fc[np.ix_([1,3,5,7,9],[1,3,5,7,9])] = np.array([
    [-(c1+c2)/m1, (c2)/m1, 0, 0, 0],
    [ (c2)/m2, -(c2+c3)/m2, (c3)/m2, 0, 0],
    [0,(c3)/m3, -(c3+c4)/m3,(c4)/m3,0],
    [0,0,(c4)/m4, -(c4+c5)/m4,(c5)/m4],
    [0,0,0,(c5)/m5, -(c5)/m5]
])

# kinematics
Fc[np.ix_([0,2,4,6,8],[1,3,5,7,9])] = np.eye(5)

# disturbance on 5th mass
Gc = np.zeros((nx,1))
Gc[9,0] = 1/m5

# position msmts on 1st and 5th masses
i1 = 0
Hq1 = np.zeros((1,nx))
Hq1[0,i1] = 1        # q1
i5 = 8
Hq5 = np.zeros((1,nx))
Hq5[0,i5] = 1        # q5
Hc = np.vstack((Hq1, Hq5))

# discrete time model
dt = 0.05
F, G, _, _, _ = cont2discrete((Fc, Gc, Hc, 0), dt)

# ------------------------------------------------------------
# Simulate system
# ------------------------------------------------------------
t = np.arange(0,50+dt,dt)
nk = len(t)

x0 = np.zeros(nx)
x0[8] = 1

np.random.seed(101)
Qsim = 0.01
w = np.sqrt(Qsim)*np.random.randn(1,nk)

x_true = np.zeros((nx,nk))
x_no_w = np.zeros((nx,nk))
x_true[:,0] = x0
x_no_w[:,0] = x0

for k in range(nk-1):
    x_no_w[:,k+1] = F @ x_no_w[:,k]
    x_true[:,k+1] = F @ x_true[:,k] + G @ w[:,k]

# measurements
Rq1 = 0.001
Rq5 = 0.001
z_q1 = (Hq1 @ x_true + np.sqrt(Rq1)*np.random.randn(1,nk)).ravel() # scalar msmt
z_q5 = (Hq5 @ x_true  + np.sqrt(Rq5)*np.random.randn(1,nk)).ravel() # scalar msmt


# ------------------------------------------------------------
# Open-loop plots
# ------------------------------------------------------------
fig, axs = plt.subplots(1,2,figsize=(16,6))

plot_openloop(t, x_true[i1,:], z=z_q1,
              x_no_w=x_no_w[i1,:], ax=axs[0])
axs[0].set_ylabel("position of 1st mass")

plot_openloop(t, x_true[i5,:], z=z_q5,
              x_no_w=x_no_w[i5,:], ax=axs[1])
axs[1].set_ylabel("position of 5th mass")

fig.suptitle("Simulated Response: 1st and 5th Masses",
             fontweight='bold')

plt.tight_layout()


# ------------------------------------------------------------
# (a) KF using z_q1
# ------------------------------------------------------------
# uses: Hq1, Rq1, z_q1
#

## YOUR CODE HERE


fig, axs = plt.subplots(1,2,figsize=(16,6))

plot_estimator(t, xhatu[i1,:], Pu[i1,i1,:],
               x_true[i1,:], plot_type='error',
               z=z_q1, ax=axs[0])
axs[0].set_ylim([-0.3,0.3])
axs[0].set_ylabel("error q1")

plot_estimator(t, xhatu[i5,:], Pu[i5,i5,:],
               x_true[i5,:], plot_type='error',
               ax=axs[1])
axs[1].set_ylim([-2,2])
axs[1].set_ylabel("error q5")

fig.suptitle("(a) Measurement z_q1", fontweight='bold')
plt.tight_layout()

# ------------------------------------------------------------
# (a) KF using z_q5
# ------------------------------------------------------------
# uses: Hq5, Rq5, z_q5
#

## YOUR CODE HERE

fig, axs = plt.subplots(1,2,figsize=(16,6))

plot_estimator(t, xhatu[i1,:], Pu[i1,i1,:],
               x_true[i1,:], plot_type='error',
               ax=axs[0])
axs[0].set_ylim([-1,1])
axs[0].set_ylabel("error q1")

plot_estimator(t, xhatu[i5,:], Pu[i5,i5,:],
               x_true[i5,:], plot_type='error',
               z=z_q5, ax=axs[1])
axs[1].set_ylim([-0.3,0.3])
axs[1].set_ylabel("error q5")

fig.suptitle("(a) Measurement z_q5", fontweight='bold')
plt.tight_layout()


# ------------------------------------------------------------
# (b) Predicted vs Updated covariance
# ------------------------------------------------------------
# uses: Hq5, Rq5, z_q5
#       solve_discrete_are
#

## YOUR CODE HERE

plt.figure(figsize=(8,6))
plt.semilogy(t, Pp[i5,i5,:], 'b:', label="predicted")
plt.semilogy(t, Pu[i5,i5,:], 'r-', label="updated")
plt.ylim([1E-4,1E-3])
plt.grid(True)
plt.xlabel("time (sec)")
plt.ylabel("estimate")
plt.legend()
plt.title("(b) Predicted vs Updated")

# ------------------------------------------------------------
# (b) Steady State error covariance
# ------------------------------------------------------------

## YOUR CODE HERE

print("Simulated measurement covariance at the final time is:")
print(Pz_sim)
print("Steady-state measurement covariance is:")
print(Hq5 @ Pss @ Hq5.T)


# ------------------------------------------------------------
# (d) Constant gain KF
# ------------------------------------------------------------
# uses: Hq5, Rq5, z_q5
#       solve_discrete_are
#


fig, axs = plt.subplots(1,2,figsize=(16,6))

plot_estimator(t, xhatu[i1,:], Pu[i1,i1,:],
               x_true[i1,:], plot_type='error',
               ax=axs[0])
axs[0].set_ylim([-0.1,0.1])
axs[0].set_ylabel("error q1")

plot_estimator(t, xhatu[i5,:], Pu[i5,i5,:],
               x_true[i5,:], plot_type='error',
               z=z_q5, ax=axs[1])
axs[1].set_ylim([-0.2,0.2])
axs[1].set_ylabel("error q5")

fig.suptitle("(d) Constant KF Gain with msmt ", fontweight='bold')
plt.tight_layout()

plt.show()